overlord/snapstate/sequence/sequence.go

```diff
@@ -0,0 +1,171 @@
+	return &RevisionSideState{Snap: snapSideInfo, Components: compSideInfo}
+}
+
+// SnapSequence is a container for a slice containing revisions of
+// snaps plus related components.
+// TODO add methods to access Revisions (length, copy, append) and
+// use them in handlers.go and snapstate.go.
+type SnapSequence struct {
+	// Revisions contains information for a snap revision and
+	// components SideInfo.
+	Revisions []*RevisionSideState
+}
+
+// MarshalJSON implements the json.Marshaler interface. We override the default
+// so serialization of the SnapState.Sequence field is compatible to what was
+// produced when it was defined as a []*snap.SideInfo. This is also the reason
+// to have SnapSequence.UnmarshalJSON and MarshalJSON/UnmarshalJSON for
+// RevisionSideState.
+func (snapSeq SnapSequence) MarshalJSON() ([]byte, error) {
+	return json.Marshal(snapSeq.Revisions)
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface
+func (snapSeq *SnapSequence) UnmarshalJSON(in []byte) error {
+	aux := []*RevisionSideState{}
+	if err := json.Unmarshal(in, &aux); err != nil {
+		return err
+	}
+	snapSeq.Revisions = aux
+	return nil
+}
+
+// SideInfos returns a slice with all the SideInfos for the snap sequence.
+func (snapSeq SnapSequence) SideInfos() []*snap.SideInfo {
+	sis := make([]*snap.SideInfo, len(snapSeq.Revisions))
+	for i, rev := range snapSeq.Revisions {
+		sis[i] = rev.Snap
+	}
+	return sis
+}
+
+// LastIndex returns the last index of the given revision in snapSeq,
+// or -1 if the revision was not found.
+func (snapSeq *SnapSequence) LastIndex(revision snap.Revision) int {
+	for i := len(snapSeq.Revisions) - 1; i >= 0; i-- {
+		if snapSeq.Revisions[i].Snap.Revision == revision {
+			return i
+		}
+	}
+	return -1
+}
+
+var ErrSnapRevNotInSequence = errors.New("snap is not in the sequence")
+
+// AddComponentForRevision adds a component to the last instance of snapRev in
+// the sequence.
+func (snapSeq *SnapSequence) AddComponentForRevision(snapRev snap.Revision, csi *snap.ComponentSideInfo) error {
+	snapIdx := snapSeq.LastIndex(snapRev)
+	if snapIdx == -1 {
+		return ErrSnapRevNotInSequence
+	}
+	revSt := snapSeq.Revisions[snapIdx]
+
+	if revSt.FindComponent(csi.Component) != nil {
+		// Component already present
+		return nil
+	}
+
+	// Append new component to components of the current snap
+	revSt.Components = append(revSt.Components, csi)
+	return nil
+}
+
+// RemoveComponentForRevision removes the cref component for the last instance
+// of snapRev in the sequence and returns a pointer to it, which might be nil
+// if not found.
+func (snapSeq *SnapSequence) RemoveComponentForRevision(snapRev snap.Revision, cref naming.ComponentRef) (unlinkedComp *snap.ComponentSideInfo) {
+	snapIdx := snapSeq.LastIndex(snapRev)
+	if snapIdx == -1 {
+		return nil
+	}
+
+	revSt := snapSeq.Revisions[snapIdx]
+	leftComp := []*snap.ComponentSideInfo{}
```

Comment:
nitpick: maybe this could be nil instead so we don't make and assign an empty slice (which could then be unnecessarily serialised)?
```suggestion
	var leftComp []*snap.ComponentSideInfo
```

Reply:
Good point, changed